Issue with GPU build using latest NVHPC compilers

## Context
cmake build with GPU fails (Debian 11, cuda 11.8, nvhpc&mpi 23.7, cmake 3.18.4) due to incompatible redefinition.
This is a fresh build starting from a Google VM (Google, Deep Learning VM with CUDA 11.8, M111, Debian 11, Python 3.10. With CUDA 11.8 preinstalled) + nvhpc + python dependencies.

### Overview of the issue
Debian 11 w/ cuda 11.8, nvhpc 23.7, cmake 3.18.4 fails.

```
cmake ..   -DNRN_ENABLE_CORENEURON=ON   -DCORENRN_ENABLE_GPU=ON   -DNRN_ENABLE_INTERVIEWS=OFF   -DNRN_ENABLE_RX3D=OFF   -DCMAKE_INSTALL_PREFIX=$HOME/install   -DCMAKE_C_COMPILER=nvc   -DCMAKE_CUDA_COMPILER=nvcc   -DCMAKE_CXX_COMPILER=nvc++
``` 
succeeds

however, when running 
```
cmake --build . --parallel
```

```
"/usr/include/stdint.h", line 127: error: incompatible redefinition of macro "UINT8_MAX" (declared at line 78 of "src/nrniv/modlunit_generated/lex.cpp")
  # define UINT8_MAX            (255)
           ^

"/usr/include/stdint.h", line 128: error: incompatible redefinition of macro "UINT16_MAX" (declared at line 81 of "src/nrniv/modlunit_generated/lex.cpp")
  # define UINT16_MAX           (65535)
           ^

"/usr/include/stdint.h", line 227: error: incompatible redefinition of macro "SIZE_MAX" (declared at line 88 of "src/nrniv/modlunit_generated/lex.cpp")
  #  define SIZE_MAX            (18446744073709551615UL)
```
  
### Expected result/behavior

generation of binaries

### NEURON setup
 
 - Version: 8.2.3
 - Installation method: 
 cmake w/ GPU
 - OS + Version: 
 Debian 11
 - Compiler + Version: 
 nvcc 11.8, nvc 23.7, nvcc 23.7
 
## Minimal working example - MWE
from Debian 11 w/ nvcc 11.8, nvc 23.7, nvcc 23.7

from the build per tutorial steps>
```
cmake ..   -DNRN_ENABLE_CORENEURON=ON   -DCORENRN_ENABLE_GPU=ON   -DNRN_ENABLE_INTERVIEWS=OFF   -DNRN_ENABLE_RX3D=OFF   -DCMAKE_INSTALL_PREFIX=$HOME/install   -DCMAKE_C_COMPILER=nvc   -DCMAKE_CUDA_COMPILER=nvcc   -DCMAKE_CXX_COMPILER=nvc++
```

then fails on>
```
cmake --build . --parallel
```
## Logs
```
"/usr/include/stdint.h", line 127: error: incompatible redefinition of macro "UINT8_MAX" (declared at line 78 of "src/nrniv/nocmodl_generated/lex.cpp")
  # define UINT8_MAX            (255)
           ^

"/usr/include/stdint.h", line 128: error: incompatible redefinition of macro "UINT16_MAX" (declared at line 81 of "src/nrniv/nocmodl_generated/lex.cpp")
  # define UINT16_MAX           (65535)
           ^

"/usr/include/stdint.h", line 227: error: incompatible redefinition of macro "SIZE_MAX" (declared at line 88 of "src/nrniv/nocmodl_generated/lex.cpp")
  #  define SIZE_MAX            (18446744073709551615UL)
            ^
 ```

Hi @pramodk 
checkout of the branch resolves the incompatible redefinition issue: however, the build fails at a (much) later point->
`cmake --build . --parallel 8`
```
NVC++-S-0053-Illegal use of void type (/home/jchen/nrn/src/coreneuron/utils/randoms/nrnran123.cpp)
NVC++/x86-64 Linux 23.7-0: compilation completed with severe errors
gmake[2]: *** [src/coreneuron/CMakeFiles/coreneuron-core.dir/build.make:849: src/coreneuron/CMakeFiles/coreneuron-core.dir/utils/randoms/nrnran123.cpp.o] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:3506: src/coreneuron/CMakeFiles/coreneuron-core.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

I do have the ability to grant access to the compute instance if given email address.

Many thanks,

James

--- a/cmake/CompilerHelper.cmake
+++ b/cmake/CompilerHelper.cmake
@@ -16,6 +16,15 @@ if(CMAKE_C_COMPILER_ID MATCHES "PGI" OR CMAKE_C_COMPILER_ID MATCHES "NVHPC")
   if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.20" AND ${CMAKE_VERSION} VERSION_LESS "3.20.3")
     string(APPEND CMAKE_DEPFILE_FLAGS_CXX "-MD<DEP_FILE>")
   endif()
+
+  # CMake versions <3.19 used to add -A when using NVHPC/PGI, which makes the compiler excessively
+  # pedantic. See https://gitlab.kitware.com/cmake/cmake/-/issues/20997. Also, stdinit.h include
+  # behaviour is different with -A (ANSI C++) and result into an error mentioned in
+  # https://github.com/neuronsimulator/nrn/issues/2563
+  if(CMAKE_VERSION VERSION_LESS 3.19)
+    list(REMOVE_ITEM CMAKE_CXX17_STANDARD_COMPILE_OPTION -A)
+  endif()
+
   if(${CMAKE_C_COMPILER_VERSION} VERSION_GREATER_EQUAL 20.7)
     # https://forums.developer.nvidia.com/t/many-all-diagnostic-numbers-increased-by-1-from-previous-values/146268/3
     # changed the numbering scheme in newer versions. The following list is from a clean start 16

--- a/src/coreneuron/CMakeLists.txt
+++ b/src/coreneuron/CMakeLists.txt
@@ -180,11 +180,6 @@ if(CORENRN_HAVE_NVHPC_COMPILER)
     # problem. If GPU support is disabled, we define R123_USE_INTRIN_H=0 to avoid the problem.
     list(APPEND CORENRN_COMPILE_DEFS R123_USE_INTRIN_H=0)
   endif()
-  # CMake versions <3.19 used to add -A when using NVHPC/PGI, which makes the compiler excessively
-  # pedantic. See https://gitlab.kitware.com/cmake/cmake/-/issues/20997.
-  if(CMAKE_VERSION VERSION_LESS 3.19)
-    list(REMOVE_ITEM CMAKE_CXX17_STANDARD_COMPILE_OPTION -A)
-  endif()
 endif()
 
 if(CORENRN_ENABLE_SHARED)

